### PR TITLE
Added GA4 Reporting

### PR DESCRIPTION
**Pull request by davidlinke**
Additionally made it so clicking the add to cart button won't trigger an event if it has the class 'disabled' or 'isDisabled’, and also improved handling of missing quantities for 'add to cart' and 'remove from cart' events

### CODE REVIEW

**Comment by davidlinke** [comment]
The following GA4 events were added:
- view_item_list
- view_item
- add_to_cart
- view_cart
- remove_from_cart
- begin_checkout
- add_shipping_info
- add_payment_info
- purchase

File: cartridges/plugin_gtm/cartridge/templates/default/gtm/gtmScript.isml [hunk 1]
```diff
@@ -37,6 +58,14 @@
         return -1;
     }
 
+    function getGA4OrderId(dataLayer) {
+        if ('ecommerce' in dataLayer && 'transaction_id' in dataLayer.ecommerce) {
+            return dataLayer.transaction_id;
```
Comment (asecaida) [inline cartridges/plugin_gtm/cartridge/templates/default/gtm/gtmScript.isml:63]: Is this right?  Conditional makes it seem that it should be `dataLayer.ecommerce.transaction_id`

File: cartridges/plugin_gtm/cartridge/controllers/GTM.js [hunk 2]
```diff
@@ -6,9 +6,9 @@ var gtmHelpers = require('*/cartridge/scripts/gtm/gtmHelpers');
 
 server.get('CustomerData', function (req, res, next) {
     var customerData = gtmHelpers.getCustomerData(req);
-    
+
     res.render('/gtm/gtmCustomerData', {
-        customerData: JSON.stringify(customerData)
+        customerData: JSON.stringify(customerData),
```
Comment (asecaida) [inline cartridges/plugin_gtm/cartridge/controllers/GTM.js:11]: Can you remove all these trailing commas please.

File: cartridges/plugin_gtm/cartridge/scripts/gtm/gtmHelpers.js [hunk 3]
```diff
@@ -47,7 +48,7 @@ function getCustomerData(res) {
  */
 function getHomeData() {
     var obj = {
-        'event': 'home'
```
Comment (asecaida) [inline cartridges/plugin_gtm/cartridge/scripts/gtm/gtmHelpers.js:50]: let's not remove the single quotes on properties.

**Review by asecaida — CHANGES_REQUESTED** [review]
I think your linter is making a lot of changes I would rather not have made, trailing commas, removing single commas on properties of object.

File: cartridges/plugin_gtm/cartridge/templates/default/gtm/gtmScript.isml (hunk 1 above)
Comment (davidlinke) [reply cartridges/plugin_gtm/cartridge/templates/default/gtm/gtmScript.isml]: Good catch, this has been fixed

File: cartridges/plugin_gtm/cartridge/controllers/GTM.js (hunk 2 above)
Comment (davidlinke) [reply cartridges/plugin_gtm/cartridge/controllers/GTM.js]: I undid all the code formatting/linting

File: cartridges/plugin_gtm/cartridge/scripts/gtm/gtmHelpers.js (hunk 3 above)
Comment (davidlinke) [reply cartridges/plugin_gtm/cartridge/scripts/gtm/gtmHelpers.js]: I undid all the code formatting/linting

File: cartridges/plugin_gtm/cartridge/scripts/gtm/gtmHelpers.js [hunk 4]
```diff
@@ -276,50 +432,124 @@ function getConfirmationData(res, step) {
...
-        case 'CheckoutServices-PlaceOrder':
-            return module.exports.getCheckoutData(6);
-        case 'Order-Confirm':
-            return module.exports.getConfirmationData(res, 7);
-    };
+function getDataLayer(res, ga4) {
+    if (ga4) {
+        // GA4 Events
+        switch (res.action) {
+            case 'Product-Show':
```
Comment (asecaida) [inline cartridges/plugin_gtm/cartridge/scripts/gtm/gtmHelpers.js:484]: Missing the `Product-ShowInCategory` case, this is a newer route in SFRA 5.x, but it should be there in both GA4 and Universal Analytics.

File: cartridges/plugin_gtm/cartridge/scripts/gtm/gtmHelpers.js [hunk 5]
```diff
@@ -276,50 +432,124 @@ function getConfirmationData(res, step) {
...
+                return false;
+        }
+    } else {
+        // Universal GA Events
+        switch (res.action) {
+            case SITE_NAME:
+            case 'Home-Show':
+            case 'Default-Start':
+                return getHomeData();
+            case 'Product-Show':
```
Comment (asecaida) [inline cartridges/plugin_gtm/cartridge/scripts/gtm/gtmHelpers.js:508]: Missing the `Product-ShowInCategory` case.

File: cartridges/plugin_gtm/cartridge/scripts/gtm/gtmHelpers.js [hunk 6]
```diff
@@ -276,50 +432,124 @@ function getConfirmationData(res, step) {
...
+    getGA4ConfirmationData: getGA4ConfirmationData,
     getConfirmationActionFieldObject: getConfirmationActionFieldObject,
     getProductArrayFromList: getProductArrayFromList,
     getSearchProducts: getSearchProducts,
     getOrderProductObject: getOrderProductObject,
-    getCheckoutData: getCheckoutData
-}
+    getGA4OrderProductObject: getGA4OrderProductObject,
+    getCheckoutData: getCheckoutData,
+    getGA4CheckoutData: getGA4CheckoutData,
```
Comment (asecaida) [inline cartridges/plugin_gtm/cartridge/scripts/gtm/gtmHelpers.js:554]: remove trailing comma

File: cartridges/plugin_gtm/cartridge/scripts/hooks/gtm/gtmhooks.js [hunk 7]
```diff
@@ -8,14 +8,32 @@ var gtmHelpers = require('*/cartridge/scripts/gtm/gtmHelpers');
...
+
+        velocity.render(
+            "$velocity.remoteInclude('GTM-HtmlHead', 'action', $action, 'datalayer', $datalayer, 'ga4datalayer', $ga4datalayer, 'gtmEnabled', $gtmEnabled, 'gtmGA4Enabled', $gtmGA4Enabled)",
+            {
+                velocity: velocity,
+                action: pdict.action,
+                datalayer: datalayer ? JSON.stringify(datalayer) : false,
+                ga4datalayer: ga4datalayer ? JSON.stringify(ga4datalayer) : false,
+                gtmEnabled: gtmEnabled,
+                gtmGA4Enabled: gtmGA4Enabled,
```
Comment (asecaida) [inline cartridges/plugin_gtm/cartridge/scripts/hooks/gtm/gtmhooks.js:33]: remove trailing comma

File: cartridges/plugin_gtm/cartridge/static/default/js/tagmanager.js [hunk 8]
```diff
@@ -9,37 +9,61 @@ var events = {
...
+            });
+        });
+
+        // Update GTM data attribute
+        $('body').on('product:updateAddToCart', function (e, response) {
+            $('button.add-to-cart, button.add-to-cart-global', response.$productContainer)
+                .attr('data-gtmdata', JSON.stringify(response.product.gtmData))
+                .attr('data-gtmga4data', JSON.stringify(response.product.gtmGA4Data));
+        });
+    },
```
Comment (asecaida) [inline cartridges/plugin_gtm/cartridge/static/default/js/tagmanager.js:66]: remove trailing comma

File: cartridges/plugin_gtm/cartridge/scripts/gtm/gtmHelpers.js [hunk 9]
```diff
@@ -153,12 +222,36 @@ function getSearchImpressionData(res) {
...
+/**
+ * @param {Object} res - current route response object
+ * @return {Object} Object containing search impression data.
+ */
+function getGA4SearchImpressionData(res) {
+    var obj = {
+        'event': 'view_item_list',
+        'ecommerce': {
+            'items': module.exports.getProductArrayFromList(module.exports.getSearchProducts(res).iterator(), module.exports.getGA4ProductObject, true)
+        },
```
Comment (asecaida) [inline cartridges/plugin_gtm/cartridge/scripts/gtm/gtmHelpers.js:240]: remove trailing comma

**Review by asecaida — CHANGES_REQUESTED** [review]

**Comment by davidlinke** [comment]
Added the Product-ShowInCategory route and remove those stray trailing commas

**Comment by asecaida** [comment]
@davidlinke thanks!  I am going to leave it open for now but let me know how testing goes with VTB.

**Pull request merged**